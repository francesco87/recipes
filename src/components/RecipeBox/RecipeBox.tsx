import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import { useNavigate } from 'react-router';
import Favorite from '../Favorite/Favorite.tsx';

export type RecipeType = {
  idMeal: string;
  strMeal: string;
  strMealAlternate: string | null;
  strCategory: string;
  strArea: string;
  strInstructions: string;
  strMealThumb: string;
  strTags: string;
  strYoutube: string;
  strIngredient1: string;
  strIngredient2: string;
  strIngredient3: string;
  strIngredient4: string;
  strIngredient5: string;
  strIngredient6: string;
  strIngredient7: string;
  strIngredient8: string;
  strIngredient9: string;
  strIngredient10: string;
  strIngredient11: string;
  strIngredient12: string;
  strIngredient13: string;
  strIngredient14: string;
  strIngredient15: string;
  strIngredient16: string;
  strIngredient17: string;
  strIngredient18: string;
  strIngredient19: string;
  strIngredient20: string;
  strMeasure1: string;
  strMeasure2: string;
  strMeasure3: string;
  strMeasure4: string;
  strMeasure5: string;
  strMeasure6: string;
  strMeasure7: string;
  strMeasure8: string;
  strMeasure9: string;
  strMeasure10: string;
  strMeasure11: string;
  strMeasure12: string;
  strMeasure13: string;
  strMeasure14: string;
  strMeasure15: string;
  strMeasure16: string;
  strMeasure17: string;
  strMeasure18: string;
  strMeasure19: string;
  strMeasure20: string;
  strSource: string;
  strImageSource: string | null;
  strCreativeCommonsConfirmed: string | null;
  dateModified: string | null;
};

type Props = {
  recipe: RecipeType;
};

const RecipeBox = ({ recipe }: Props) => {
  const navigate = useNavigate();

  const onClick = () => {
    localStorage.setItem('currentRecipe', JSON.stringify(recipe));
    navigate(`/recipe/${recipe.idMeal}`);
  };

  return (
    <Card sx={{ width: '100%', position: 'relative' }}>
      <CardMedia component="img" alt={recipe.strMeal} height="180" image={recipe.strMealThumb} />
      <CardContent>
        <Favorite id={recipe.idMeal} />
        <Typography
          variant="h6"
          component="div"
          sx={{
            height: '65px',
            display: '-webkit-box',
            WebkitBoxOrient: 'vertical',
            WebkitLineClamp: 2,
            overflow: 'hidden',
          }}
        >
          {recipe.strMeal}
        </Typography>
        <Typography
          variant="body2"
          sx={{
            color: 'text.secondary',
            display: '-webkit-box',
            WebkitBoxOrient: 'vertical',
            WebkitLineClamp: 2,
            overflow: 'hidden',
          }}
        >
          {recipe.strInstructions}
        </Typography>
      </CardContent>
      <CardActions>
        <Button size="small" onClick={onClick}>
          See details...
        </Button>
      </CardActions>
    </Card>
  );
};

export default RecipeBox;
