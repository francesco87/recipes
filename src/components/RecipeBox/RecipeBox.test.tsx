import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import RecipeBox from './RecipeBox';
import { BrowserRouter } from 'react-router';
import recipeMock from '../../api/mock/RecipeMock.ts';

const mockNavigate = vi.fn();

vi.mock('@tanstack/react-router', () => ({
  useNavigate: () => mockNavigate,
}));

vi.mock('../Favorite/Favorite.tsx', () => ({
  default: ({ id }: { id: string }) => <div data-testid="favorite-mock">Favorite Mock {id}</div>,
}));

// Setup localStorage mock
beforeEach(() => {
  mockNavigate.mockClear();
  vi.stubGlobal('localStorage', {
    setItem: vi.fn(),
    getItem: vi.fn(),
    removeItem: vi.fn(),
    clear: vi.fn(),
  });
});

describe('RecipeBox Component', () => {
  const renderComponent = () =>
    render(
      <BrowserRouter>
        <RecipeBox recipe={recipeMock} />
      </BrowserRouter>
    );

  it('Show recipe title and description', () => {
    renderComponent();
    expect(screen.getByText('Pizza Margherita')).toBeInTheDocument();
    expect(screen.getByText('Bake it in the oven.')).toBeInTheDocument();
  });

  it('Show recipe image', () => {
    renderComponent();
    const img = screen.getByRole('img') as HTMLImageElement;
    expect(img).toHaveAttribute('src', recipeMock.strMealThumb);
    expect(img).toHaveAttribute('alt', recipeMock.strMeal);
  });

  it('Show favorite recipe', () => {
    renderComponent();
    expect(screen.getByTestId('favorite-mock')).toBeInTheDocument();
    expect(screen.getByTestId('favorite-mock')).toHaveTextContent('12345');
  });

  it('RecipeBox Snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
