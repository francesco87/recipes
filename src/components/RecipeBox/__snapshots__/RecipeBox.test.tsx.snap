// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecipeBox Component > RecipeBox Snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-42nkya-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <img
      alt="Pizza Margherita"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1kuwa16-MuiCardMedia-root"
      height="180"
      src="https://example.com/pizza.jpg"
    />
    <div
      class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
    >
      <div
        data-testid="favorite-mock"
      >
        Favorite Mock 
        12345
      </div>
      <div
        class="MuiTypography-root MuiTypography-h6 css-1rlo9o3-MuiTypography-root"
      >
        Pizza Margherita
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-h85k1g-MuiTypography-root"
      >
        Bake it in the oven.
      </p>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-1ccsd5i-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        See details...
      </button>
    </div>
  </div>
</div>
`;
